src/nanotron/trainer.py

```diff
@@ -299,12 +312,20 @@ def train(
         torch.cuda.empty_cache()
         with prof:
             for self.iteration_step in range(self.start_iteration_step + 1, self.config.tokens.train_steps + 1):
+                # NOTE: the second condition is for if we have a single training stages
+                # and this is the first iteration step
+                if self.config.data.dataset_stages is not None or (
+                    self.config.data.dataset_stages is None and self.iteration_step == 1
+                ):
+                    self._update_dataloader_based_on_training_stages(dataloader_or_dls)
```

Comment:
maybe to keep the `train` code simpler, we can move the if logic inside `self._update_dataloader_based_on_training_stages`